contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,282 @@
+    Token string `json:"token,omitempty"`
+    // Expiry indicates a unix time when the provided credentials expire.
+    Expiry int64 `json:"expiry,omitempty"`
+}
+
+// Response defines metadata about a failed request, including HTTP status code and
+// response headers.
+type Response struct {
+    // HTTP header returned by the server.
+    Header map[string][]string `json:"header,omitempty"`
+    // HTTP status code returned by the server.
+    Code int32 `json:"code,omitempty"`
+}
+
+// ExecInfo is structed information passed to the plugin.
+type ExecInfo struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Response is populated when the transport encounters HTTP status codes, such as 401,
+    // suggesting previous credentials were invalid.
+    // +optional
+    Response *Response `json:"response,omitempty"`
+
+    // Interactive is true when the transport detects the command is being called from an
+    // interactive prompt.
+    Interactive bool `json:"interactive,omitempty"`
+}
+```
+
+To instruct client-go to use the bearer token `BEARER_TOKEN`, a plugin would print:
+
+```terminal
+$ ./kubectl-example-auth-plugin
+{
+  "kind": "ExecCredentials",
+  "apiVersion":"client.authentication.k8s.io/v1alpha1",
+  "token":"BEARER_TOKEN"
+}
+```
+
+To surface runtime-based information to the plugin, such as a request body for request
+signing, client-go will set the environment variable `KUBERNETES_EXEC_INFO` to a JSON
+serialized Kubernetes object when calling the plugin.
+
+
+```terminal
+KUBERNETES_EXEC_INFO='{
+    "kind":"ExecInfo",
+    "apiVersion":"client.authentication.k8s.io/v1alpha1",
+    "response": {
+        "code": 401,
+        "header": {
+            "WWW-Authenticate": ["Bearer realm=\"Access to the staging site\""]
+        }
+    },
+    "interactive": true
+}'
+```
+
+### Caching
+
+kubectl repeatedly [re-initializes transports](https://github.com/kubernetes/kubernetes/issues/37876)
+while client-go transports are long lived over many requests. As a result naive auth
+provider implementations that re-request credentials on every request have historically
+been slow.
+
+Plugins will be called on client-go initialization, and again when the API server returns
+a 401 HTTP status code indicating expired credentials. Plugins can indicate their credentials
+explicit expiry using the `Expiry` field on the returned `ExecCredentials` object, otherwise
+credentials will be cached throughout the lifetime of a program.
+
+## Kubeconfig changes
+
+The current `AuthProviderConfig` uses `map[string]string` for configuration, which
+makes it hard to express things like a list of arguments or list key/value environment
+variables. As such, `AuthInfo` should add another field which expresses the `exec`
+config. This has the benefit of a more natural structure, but the trade-off of not being
+compatible with the existing `kubectl config set-credentials` implementation.
```

Comment:
Is there any way today to use `kubectl` to write below config into `~/.kube/config`?